File: 算法/Java/第3章 排序算法/3.8 桶排序/BucketSort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Collections;

public class BucketSort {
    public static void main(String[] args) {
        int[] arr = {4, 6, 23, 3, 21, 8, 41, 10};
        bucketSort(arr);
        System.out.println(Arrays.toString(arr));
    }

    public static void bucketSort(int[] arr) {
        int max = arr[0];
        int min = arr[0];
        for(int i = 1; i < arr.length; i++) {
            if(arr[i] > max) {
                max = arr[i];
            }
            if(arr[i] < min) {
                min = arr[i];
            }
        }
        int diff = max - min;

        // 创建桶
        int bucketNum = arr.length;
        List<LinkedList<Integer>> buckets = new ArrayList<>(bucketNum);
        for(int i = 0; i < bucketNum; i++) {
            buckets.add(new LinkedList<>());
        }

        // 遍历数组，把元素放入对应桶中
        for(int i = 0; i < arr.length; i++) {
            // 计算当前元素所放置的桶号
            int num = (arr[i] - min) / (diff / (bucketNum - 1));
            buckets.get(num).add(arr[i]);
        }

        // 桶内排序
        for(int i = 0; i < bucketNum; i++) {
            Collections.sort(buckets.get(i));
        }

        // 取出元素
        int cnt = 0;
        for(LinkedList<Integer> bucket : buckets) {
            for(int data : bucket) {
                arr[cnt++] = data;
            }
        }
    }
}

File: 算法/Python/第3章 排序算法/3.8 桶排序/bucket_sort.py
def bucket_sort(lst):
    n = len(lst)
    max_elem = max(lst)
    min_elem = min(lst)
    diff = max_elem - min_elem

    # 创建桶
    bucket_num = n
    buckets = [[] for _ in range(bucket_num)]

    # 遍历列表，把元素放入对应桶中
    for i in range(n):
        # 计算当前元素所放置的桶号
        num = (lst[i] - min_elem) // (diff // (bucket_num - 1))
        buckets[num].append(lst[i])

    # 桶内排序
    for i in range(bucket_num):
        buckets[i].sort()
    
    # 取出元素
    cnt = 0
    for bucket in buckets:
        for data in bucket:
            lst[cnt] = data
            cnt += 1

def main():
    lst = [4, 6, 23, 3, 21, 8, 41, 10]
    bucket_sort(lst)
    print(lst)

if __name__ == "__main__":
    main()